src/smc-webapp/editor.coffee

```diff
@@ -1282,42 +1302,45 @@ class CodeMirrorEditor extends FileEditor
                 (cb) =>
                     @save(cb)
                 (cb) =>
-                    salvus_client.print_to_pdf
+                    options =
+                        title      : dialog.find(".salvus-file-print-title").text()
+                        author     : dialog.find(".salvus-file-print-author").text()
+                        date       : dialog.find(".salvus-file-print-date").text()
+                        contents   : dialog.find(".salvus-file-print-contents").is(":checked")
+                        subdir     : is_subdir
+                        extra_data : misc.to_json(@syncdoc.print_to_pdf_data())  # avoid de/re-json'ing
+
+                    printing.Printer(@).print
                         project_id  : @project_id
                         path        : @filename
-                        options     :
-                            title      : dialog.find(".salvus-file-print-title").text()
-                            author     : dialog.find(".salvus-file-print-author").text()
-                            date       : dialog.find(".salvus-file-print-date").text()
-                            contents   : dialog.find(".salvus-file-print-contents").is(":checked")
-                            subdir     : is_subdir
-                            extra_data : misc.to_json(@syncdoc.print_to_pdf_data())  # avoid de/re-json'ing
+                        options     : options
                         cb          : (err, _pdf) =>
                             if err
                                 cb(err)
                             else
                                 pdf = _pdf
                                 cb()
                 (cb) =>
-                    salvus_client.read_file_from_project
-                        project_id : @project_id
-                        path       : pdf
-                        cb         : (err, mesg) =>
+                    redux.getProjectActions(@project_id).file_nonzero
```

Comment:
Actions are not supposed to return anything.   They're supposed to set something in motion, whose impact is only via changes to the store.  There are a few critical cases (e.g., private internal methods) with callbacks.   Is there any way you can do whatever file_nonzero does but without a callback or move it somewhere that isn't an action?

ACTIONS:  take input and modify the store.  Do not return values. 

STORE: Functions/props compute stuff about the store, but never change the store. 

The Actions vs store business follows this design pattern:  http://martinfowler.com/bliki/CQRS.html